durable-storage/src/database.rs

```diff
@@ -0,0 +1,125 @@
+// SPDX-FileCopyrightText: 2025 Trilitech <contact@trili.tech>
+//
+// SPDX-License-Identifier: MIT
+
+use std::sync::Arc;
+
+use crate::merkle_layer::Key;
+use crate::merkle_layer::MerkleLayer;
+use crate::merkle_layer::MerkleLayerError;
+use crate::persistence_layer::PersistenceLayer;
+use crate::persistence_layer::PersistenceLayerError;
+
+#[derive(Clone)]
+#[expect(dead_code, reason = "Implemented in RV-827")]
+pub struct Database {
+    persistent: Arc<PersistenceLayer>,
+    merkle: Arc<MerkleLayer>,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum DatabaseError {
+    #[error("The offset is too large")]
+    BadOffset,
+
+    #[error("The provided key is expected to exist but does not")]
+    KeyDoesNotExist,
+
+    #[error("Merkle layer error: {0}")]
+    MerkleLayer(#[from] MerkleLayerError),
+
+    #[error("Persistence layer error: {0}")]
+    PersistenceLayer(#[from] PersistenceLayerError),
+}
+
+impl Database {
+    #[expect(dead_code, reason = "Implemented in RV-827")]
+    /// Remove a key from the database.
+    ///
+    /// Fails if the key is too long.
+    pub(crate) fn delete(&mut self, key: &[u8]) -> Result<(), DatabaseError> {
+        let _key = Key::new(key)?;
```

Comment:
Let's use the `Key` type for keys. I think this isn't the right place to parse the key.

```suggestion
    pub(crate) fn delete(&mut self, key: &Key) -> Result<(), DatabaseError> {
```